Online DB configuration objects

lmoureaux opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

This PR is to track my progress on DB configuration objects (#213).

Currently implemented:
* Python script to generate `xsd` templates and class definitions from a list of fields.
* Integration of the script into the `Makefile` for the VFAT3 chip configuration.
* XMLSerializationData class representing the structure of XML documents in C++
* Function to turn XMLSerializationData into a DOM tree
* Function that fills an XMLSerializationData object with the contents of a DOM tree
* Tests of serialization functionality:
     * Create a DOM tree and check that it isn't empty
     * Create a DOM tree and validate it against the schema
     * Create a DOM tree, load it and check that we get back the initial object

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue(s) here. -->
#213, #214 and #215 for VFAT3 chip and channel configuration data.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
* Checked that a very basic code importing the generated class compiles
* The `xml` file in the repo can be checked using the generated `xsd` template
* Tests compile and yield the expected results
* The tests have been run under Valgrind and no leaks are found 

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly. [Link to documentation](http://homepage.iihe.ac.be/~lmoureau/GEM/cmsgemos/doc/build/html/md_gemonlinedb_gemonlinedb.html)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

<!--- Template thanks to https://www.talater.com/open-source-templates/#/page/99 -->

jsturdy commented on .gitignore:
```diff
@@ -46,6 +46,15 @@ __init__.py
+
+# Do ignore gemonlinedb auto-generated files
+gemonlinedb/**/*Gen.[h,cpp]
+
```
It is probably more appropriate to put these `gemonlinedb` specific items in, e.g., `gemonlinedb/.gitignore`

lmoureaux commented:
I'm experiencing trouble when trying to add documentation. Namely,
```
make doc
```
generates the following output:
```
Generating doxygen
make: *** [doc] Error 1
```

Removing the `>/dev/null` in the Makefile, I get:
```
Generating doxygen
error: configuration file ./doc/cmsgemos.cfg not found!
```

Indeed the file doesn't exist. Am I doing something wrong?

mexanick commented:
Indeed, the configuration for doxygen should be present. It is a bug (eventually just missed addition to corresponding git commit). Please open an issue and assign to @jsturdy

jsturdy commented:
Fixed as per #224

lmoureaux commented:
After over 1,000 lines of code, I'm thinking of making this PR converge. So let me declare both feature and API freeze, and ask for review of the API and build system as a first pass.

Everything I tested is working as expected apart from a memory leak reported by Valgrind, whose origin is not clear. I could argue that it's small (~180B) and located in a test function... but still I don't like that. I'll try to figure it out.

I need input to tackle the following issues:

* The problem with XML templates mentioned at https://github.com/cms-gem-daq-project/cmsgemos/issues/213#issuecomment-441756369. If `_NAME`/`_VALUE` pairs are needed, this will be handled later (several new classes needed); otherwise a small patch here will be enough.
* Installing `xsd` schemas. As far as I can tell, the `Makefile` system has only support for libraries. I know it's just one `install` away, but I've god no idea where to put the files... Also, I don't have access to a system where I can test RPM packages at will.

Problems delayed to future PRs:

* Locating `xsd` schemas at run time. Their location can be either in the source tree (convenient while coding), in a locally-installed `cmsgemos` (like the standard variable `PREFIX` would do, if the build system supports something equivalent; maybe that's what `BUILD_HOME` does?) or in `/opt/cmsgemos/share` (I assume). I'll likely set up a search path and expose it through an env variable.
* Integrating tests with the CI.
* OH, AMC and AMC13 configurations.

jsturdy commented:
> The problem with XML templates mentioned at #213 (comment). If _NAME/_VALUE pairs are needed, this will be handled later (several new classes needed); otherwise a small patch here will be enough.

I still need to follow up on this and check the structure of the DB, but for now let's assume we won't have these wonky entries

> Installing `xsd` schemas. As far as I can tell, the `Makefile` system has only support for libraries. I know it's just one install away, but I've god no idea where to put the files... Also, I don't have access to a system where I can test RPM packages at will.

Have the RPM put them in `/opt/cmsgemos/share` (maybe `onlinedb/xml` subfolder)

> Locating xsd schemas at run time. Their location can be either in the source tree (convenient while coding), in a locally-installed cmsgemos (like the standard variable PREFIX would do, if the build system supports something equivalent; maybe that's what BUILD_HOME does?) or in /opt/cmsgemos/share (I assume). I'll likely set up a search path and expose it through an env variable.

"Properly" set up DAQ machines will have `CMSGEMOS_ROOT` set, so you can work from the assumption that `CMSGEMOS_ROOT/share` (as above) will be resolveable

I'll start going through the PR tomorrow.

jsturdy commented on gemonlinedb/.gitignore:
```diff
@@ -0,0 +1,9 @@
+!defs/*.json
+
+# Don't ignore common xsd definitions
+!schema/common.xsd
```
Actually, in the RPM definitions,the contents of `xml` are copied to `$RPM_BUILD_ROOT%{_prefix}/share/%{_package}/xml/`, so if you put the `schema` subdirectory there, then the RPM should already contain them

jsturdy commented on doc/cmsgemos.cfg:
```diff
@@ -588,7 +588,8 @@ WARN_LOGFILE           =
 
-INPUT                  = $(BUILD_HOME)/$(GEM_OS_PROJECT)
+INPUT                  = $(BUILD_HOME)/$(GEM_OS_PROJECT) \
+                         $(BUILD_HOME)/$(GEM_OS_PROJECT)/gemonlinedb/gemonlinedb.md
```
What is the result of this?
For the documentation, there will be two parts:
* the API (code comments generated by `doxygen`)
* the usage manual (done from the `md`/`rst` files in a nice readthedocs/sphinx style)
They will not be pushed to the same place, so keeping their creation distinct was the original goal

lmoureaux replied:
It creates [this page](http://homepage.iihe.ac.be/~lmoureau/GEM/cmsgemos/doc/build/html/md_gemonlinedb_gemonlinedb.html). I intend to use it as a technical high-level overview overview of the package (something like [this](http://doc.qt.io/qt-5/application-windows.html)), and would like to include links to detailed class documentation.

If there's a way to reference Doxygen output from Sphinx, I'd be glad to use it. For now `make doc` doesn't run Sphinx...

jsturdy replied:
OK, definitely fits the part I mentioned above.

> If there's a way to reference Doxygen output from Sphinx, I'd be glad to use it. For now make doc doesn't run Sphinx...

Not currently, but on the internal (@jsturdy's brain) to-do list for once core features are fully implemented :-)
What you can assume will exist is a site structure like:
```
cmsgemos/sw/release/Maj.Min/api/latest (symlink to latest built version of MajMin)
cmsgemos/sw/release/Maj.Min/api/Maj.Min.Patch
cmsgemos/sw/release/Maj.Min/repo/<arch>/yum-repo-structure
```

What is not yet decided is where the global "user" docs will go, wether to a `cmsgemdaq` aggregate site, or a package specific site (I slightly lean towards the latter, with the former hosting a more overview type set of info)
```
cmsgemos/doc/<sphinx/rtd style doc site, versioned?>
```
This allows the user docs to be more dynamic, while allowing full API access to the versioned `doxygen` generated info based on releases.

So if (in your spare time) you want to think on how this can all work together, please feel free,
You can then open an issue when you have a concrete proposal and we can further discuss/refine.
Until then, I'd prefer that the `doxygen` config remain unchanged

lmoureaux replied:
I'll include include top-level API docs into a docstring for `gem::onlinedb` once I have an idea of what the final API will look like.

jsturdy commented on gemonlinedb/include/gem/onlinedb/DataSet.h:
```diff
@@ -0,0 +1,103 @@
+        private:
+            friend class XMLSerializationData<ConfigurationType>;
+
+            std::string comment, version;
```
Style within `cmsgemos` is generally to make `class` variables have a `m_` prefix, to differentiate them from function local variables.
The same is not applied to `struct` `typedefs` (one can easily argue that this is inconsistent, but it is what is currently done)

lmoureaux replied:
Applied to all private member variables. Public variables and types untouched.

jsturdy replied:
> Public variables and types untouched.

Reason for not also applying to public variables?

lmoureaux replied:
Many things in cmsgemos look like Java to me, so I follow the [Java conventions](https://en.wikipedia.org/wiki/Naming_convention_(programming)#Java). In Java, public member variables and types don't take any prefix. Since private variables are hidden by definition, no convention exists and developers use whatever they like.

Moreover, I could find the following `struct`s in cmsgemos: [[1]](https://github.com/cms-gem-daq-project/cmsgemos/blob/develop/gembase/include/gem/base/GEMApplication.h#L183) [[2]](https://github.com/cms-gem-daq-project/cmsgemos/blob/develop/gembase/include/gem/base/GEMMonitor.h#L182), and none of them uses leading `m_`s. The one exception I found is [`GEMApplication`](https://github.com/cms-gem-daq-project/cmsgemos/blob/develop/gembase/include/gem/base/GEMApplication.h#L211) whose interface doesn't seem to be final anyway.

Note that in this PR, all public member variable belong to structs whose sole functionality is to group sets of related variables together. If at some point I need to add functions, I'll turn them into classes, add encapsulation and `m_` prefixes.

jsturdy replied:
OK, I agree that the utilized style is a bit of a mix between various projects I've worked on and various styles I've read through (though when starting this one I tried to apply a consistent set of them...)

> Moreover, I could find the following `struct`s in `cmsgemos`: [1] [2], and none of them uses leading `m_`s.

As mentioned in the initial comment, `structs` are explicitly treated differently, and their member elements are not prefixed with `m_` (or `p_` for a member pointer as you saw in `GEMApplication`)

> Note that in this PR, all public member variable belong to `struct`s

OK, then as per the aforementioned style, everything should be consistent, and I won't belabour the point further

jsturdy commented on gemonlinedb/include/gem/onlinedb/VFAT3ChannelConfiguration.h:
```diff
@@ -0,0 +1,133 @@
+            /**
+             * @brief The number of channels in a VFAT 3
+             */
+            static const constexpr std::size_t channelCount = 128;
```
Consts should be all caps (underscore delimited)

jsturdy commented on gemonlinedb/include/gem/onlinedb/VFAT3ChannelConfiguration.h:
```diff
@@ -0,0 +1,133 @@
+                    /**
+                     * @brief Arming comparator threshold trimming: amplitude.
+                     */
+                    std::uint16_t armDacCorrectionAmplitude : 6;
```
Mentioned to @bdorney in a different PR, adopt an "almost" camel case style, i.e., a DAC is a thing (not a `Dac` or a `dac`), so if it's the second word in a variable name, it should be all caps, if it's the first word, all lower case; for words that don't have a specific capitalization, standard camel case should be used

jsturdy commented on gemonlinedb/include/gem/onlinedb/VFAT3ChannelConfiguration.h:
```diff
@@ -0,0 +1,133 @@
+        public:
+            static const char *extTableName() { return "GEM_VFAT3_CHANNEL_CONF"; }
+            static const char *typeName() { return "GEM VFAT3 Channel Conf Lookup Table"; }
+            static const char *kindOfPart() { return "GEM VFAT3"; }
```
Is there a reason these are `char*` instead of `std::string`?
If a function needs them to be `char*` when passed, in, you can always `var.c_str()`

jsturdy commented on gemonlinedb/include/gem/onlinedb/XMLSerializationData.h:
```diff
@@ -0,0 +1,248 @@
+            /**
+             * @brief Queries the list of data sets present in @c dom
+             */
+            xercesc::DOMXPathResult *queryDataSets(const DOMDocumentPtr &dom);
```
Did you look into using `shared_ptr` or `unique_ptr` to encapsulate the `xercesc` pointer objects?
I know it's not always trivial, but I did a bunch of this with the `xdaq` core pointer objects and it helped clean up a few memory leaks.

jsturdy commented on gemonlinedb/parseDef.py:
```diff
@@ -0,0 +1,270 @@
+    assert(type(json[fieldName]) == targetType)
+    return json[fieldName]
+
+def _toCamelCase(snake_string, capitalizeFirstLetter=False):
```
I see, probably this would be where implementation of the "almost" camel case from the comment above would give the most trouble...

lmoureaux replied:
Fixed by supporting manual override of C++ names in the JSON file.

jsturdy commented on gemonlinedb/test/testXMLSerialization.cc:
```diff
@@ -0,0 +1,182 @@
+
+        // Cleanup
+        delete lsser;
+        delete implLS;
```
You shouldn't `delete` anything you didn't allocate, i.e., create with `new`.
Perhaps behind the scenes `xercesc` didn't do some cleanup correctly so it leaks here, but it's potentially problematic in that there *could* be a double free/delete problem down the road.

lmoureaux commented:
Fixed all issues. Please find some comments below.

### Naming convention

It reminds me when I was writing Java code years ago :-)

How does one write "VFAT DAC" in the naming convention? A method called `getVFATDAC` would look weird; `getVfatDac` doesn't comply to the rules, nor does `getVFAT_DAC`.

### Xerces memory management

Xerces uses old-style memory management with pointers. The following rules apply:

* Anything returned by a DOM object is owned by the implementation (reference/`weak_ptr` semantics). I use bare pointers for these and hope Xerces' exception handling is sensible.
* The user *may* be responsible for releasing memory returned by `DOMImplementation::createFoo` and `DOMDocument::createBar`. Since I couldn't find a list of methods returning managed memory, I had discover the API using trial-and-error.
* The policy for anything else is unspecified. I discover it by trial-and-error.

I now use `std::unique_ptr` when appropriate; all `delete` statements are gone.

jsturdy commented:
> Fixed all issues. Please find some comments below.

Did your remaining leak get removed?

> A method called `getVFATDAC` would look weird

To me that looks perfectly fine (and better than both of the alternatives) :-)

However, does the method/variable live inside something `VFAT` specific?  If so, removing the `VFAT` part of the function/variable name results a cleaner `getDAC` (agree, this may *only* apply to the specific case you chose to illustrate, but it can generally be extended to yield a cleaner interface)

lmoureaux commented:
>> Fixed all issues. Please find some comments below.
>
> Did your remaining leak get removed?

Yes. It was caused by an exception being thrown before deletion of a pointer. Everything is clean now that I use `std::unique_ptr`.

>> A method called getVFATDAC would look weird
>
> To me that looks perfectly fine (and better than both of the alternatives) :-)

Tastes and colors :-)

> However, does the method/variable live inside something VFAT specific? If so, removing the VFAT part of the function/variable name results a cleaner getDAC (agree, this may only apply to the specific case you chose to illustrate, but it can generally be extended to yield a cleaner interface)

Some registers have names that resemble the example above, but I created it to highlight the usual ambiguities with this convention. If at some point we have a NAS DAQ, we'll have `getNASDAQ`...

jsturdy commented on gemonlinedb/parseDef.py:
```diff
@@ -147,15 +155,15 @@ class {baseName}Gen
 }} /* namespace gem */
 
-#endif // _{baseName}Gen_h_
+#endif // GEM_ONLINEDB_DETAIL_{baseName}GEN_h
```
`_h` -> `_H`
Also, will `baseName` always be all caps?

jsturdy commented on gemonlinedb/parseDef.py:
```diff
@@ -115,8 +123,8 @@ def _makeHeader(config, className):
-#ifndef _{baseName}Gen_h_
-#define _{baseName}Gen_h_
+#ifndef GEM_ONLINEDB_DETAIL_{baseName}GEN_h
+#define GEM_ONLINEDB_DETAIL_{baseName}GEN_h
```
`_h` -> `_H`
Also, will `baseName` always be all caps?

jsturdy commented:
> If at some point we have a `NAS` `DAQ`, we'll have `getNASDAQ`

And then we profit! :-D
But yes we can try to take care when designing the inputs here as well

lmoureaux commented:
Edited the relevant commits to address new comments. Reverted the commit that modified the Doxygen configuration and introduced some cross-referencing in the Markdown file.

jsturdy commented:
Just trying to get to what is implemented here: 
* object representing XML specification of a VFAT3 chip configuration record
  * including, XML specification of a VFAT3 channel configuration record
* methods to access the individual configuration parameters/properties
* `xerces` parsing back end to make aforementioned functionality work

Is this a good executive summary?

jsturdy commented:
Here is the updated list of the different views, confirming the removal of the `_NAME`/`_VALUE` entries:

#### `AMC13_CONFIGURATION`
```
Columns for GEM_AMC13_CONFIGURATION
Executing query: select * from CMS_GEM_MUON_COND.GEM_AMC13_CONFIGURATION WHERE 1 = 0
['RECORD_ID',
'CONDITION_DATA_SET_ID',
'FED_ID',
'ENABLE_LOCALTTC',
'HOSTNAME']
```

#### `AMC_CONFIGURATION`
```
Columns for GEM_AMC_CONFIGURATION
Executing query: select * from CMS_GEM_MUON_COND.GEM_AMC_CONFIGURATION WHERE 1 = 0
['RECORD_ID',
'CONDITION_DATA_SET_ID',
'DAQ_DAV_TIMEOUT',
'DAQ_INPUT_ENABLE_MASK',
'DAQ_OH0_EOE_TIMEOUT',
'DAQ_OH1_EOE_TIMEOUT',
'DAQ_OH2_EOE_TIMEOUT',
'DAQ_OH3_EOE_TIMEOUT',
'DAQ_OH4_EOE_TIMEOUT',
'DAQ_OH5_EOE_TIMEOUT',
'DAQ_OH6_EOE_TIMEOUT',
'DAQ_OH7_EOE_TIMEOUT',
'DAQ_OH8_EOE_TIMEOUT',
'DAQ_OH9_EOE_TIMEOUT',
'DAQ_OH10_EOE_TIMEOUT',
'DAQ_OH11_EOE_TIMEOUT',
'DAQ_OH0_VFAT_MASK',
'DAQ_OH1_VFAT_MASK',
'DAQ_OH2_VFAT_MASK',
'DAQ_OH3_VFAT_MASK',
'DAQ_OH4_VFAT_MASK',
'DAQ_OH5_VFAT_MASK',
'DAQ_OH6_VFAT_MASK',
'DAQ_OH7_VFAT_MASK',
'DAQ_OH8_VFAT_MASK',
'DAQ_OH9_VFAT_MASK',
'DAQ_OH10_VFAT_MASK',
'DAQ_OH11_VFAT_MASK',
'DAQ_ZERO_SUPPRESSION_EN',
'TTC_CONFIG_CMD_BC0',
'TTC_CONFIG_CMD_CALPULSE',
'TTC_CONFIG_CMD_EC0',
'TTC_CONFIG_CMD_HARD_RESET',
'TTC_CONFIG_CMD_OC0',
'TTC_CONFIG_CMD_RESYNC',
'TTC_CONFIG_CMD_START',
'TTC_CONFIG_CMD_STOP',
'TTC_CONFIG_CMD_TEST_SYNC',
'TRIGGER_OH_KILL_MASK',
'USE_OH_V3B_MAPPING',
'EOE_USE_OH_EC_BC',
'ENABLE_INPUT_AUTOKILL',
'INPUT_AUTOKILL_THRESH']
```

#### `OHV2_CONFIGURATION`
```
Columns for GEM_OHV2_CONFIGURATION
Executing query: select * from CMS_GEM_MUON_COND.GEM_OHV2_CONFIGURATION WHERE 1 = 0
['RECORD_ID',
'CONDITION_DATA_SET_ID',
'VFAT_DROP_BAD_CRC',
'VFAT_SBIT_MASK',
'VFAT_TRK_MASK',
'HDMI_SBIT_MODE0',
'HDMI_SBIT_MODE1',
'HDMI_SBIT_MODE2',
'HDMI_SBIT_MODE3',
'HDMI_SBIT_MODE4',
'HDMI_SBIT_MODE5',
'HDMI_SBIT_SEL0',
'HDMI_SBIT_SEL1',
'HDMI_SBIT_SEL2',
'HDMI_SBIT_SEL3',
'HDMI_SBIT_SEL4',
'HDMI_SBIT_SEL5']
```

#### `OHV3_CONFIGURATION`
```
Columns for GEM_OHV3_CONFIGURATION
Executing query: select * from CMS_GEM_MUON_COND.GEM_OHV3_CONFIGURATION WHERE 1 = 0
['RECORD_ID',
'CONDITION_DATA_SET_ID',
'BXN_OFFSET',
'HDMI_SBIT_MODE0',
'HDMI_SBIT_MODE1',
'HDMI_SBIT_MODE2',
'HDMI_SBIT_MODE3',
'HDMI_SBIT_MODE4',
'HDMI_SBIT_MODE5',
'HDMI_SBIT_MODE6',
'HDMI_SBIT_MODE7',
'HDMI_SBIT_SEL0',
'HDMI_SBIT_SEL1',
'HDMI_SBIT_SEL2',
'HDMI_SBIT_SEL3',
'HDMI_SBIT_SEL4',
'HDMI_SBIT_SEL5',
'HDMI_SBIT_SEL6',
'HDMI_SBIT_SEL7',
'SOT_TAP_DELAY_VFAT0',
'SOT_TAP_DELAY_VFAT1',
'SOT_TAP_DELAY_VFAT2',
'SOT_TAP_DELAY_VFAT3',
'SOT_TAP_DELAY_VFAT4',
'SOT_TAP_DELAY_VFAT5',
'SOT_TAP_DELAY_VFAT6',
'SOT_TAP_DELAY_VFAT7',
'SOT_TAP_DELAY_VFAT8',
'SOT_TAP_DELAY_VFAT9',
'SOT_TAP_DELAY_VFAT10',
'SOT_TAP_DELAY_VFAT11',
'SOT_TAP_DELAY_VFAT12',
'SOT_TAP_DELAY_VFAT13',
'SOT_TAP_DELAY_VFAT14',
'SOT_TAP_DELAY_VFAT15',
'SOT_TAP_DELAY_VFAT16',
'SOT_TAP_DELAY_VFAT17',
'SOT_TAP_DELAY_VFAT18',
'SOT_TAP_DELAY_VFAT19',
'SOT_TAP_DELAY_VFAT20',
'SOT_TAP_DELAY_VFAT21',
'SOT_TAP_DELAY_VFAT22',
'SOT_TAP_DELAY_VFAT23',
'TRIG_CTRL_VFAT_MASK',
'TRIG_TAP_DELAY_VFAT0_BIT0',
'TRIG_TAP_DELAY_VFAT0_BIT1',
'TRIG_TAP_DELAY_VFAT0_BIT2',
'TRIG_TAP_DELAY_VFAT0_BIT3',
'TRIG_TAP_DELAY_VFAT0_BIT4',
'TRIG_TAP_DELAY_VFAT0_BIT5',
'TRIG_TAP_DELAY_VFAT0_BIT6',
'TRIG_TAP_DELAY_VFAT0_BIT7',
'TRIG_TAP_DELAY_VFAT1_BIT0',
'TRIG_TAP_DELAY_VFAT1_BIT1',
'TRIG_TAP_DELAY_VFAT1_BIT2',
'TRIG_TAP_DELAY_VFAT1_BIT3',
'TRIG_TAP_DELAY_VFAT1_BIT4',
'TRIG_TAP_DELAY_VFAT1_BIT5',
'TRIG_TAP_DELAY_VFAT1_BIT6',
'TRIG_TAP_DELAY_VFAT1_BIT7',
'TRIG_TAP_DELAY_VFAT2_BIT0',
'TRIG_TAP_DELAY_VFAT2_BIT1',
'TRIG_TAP_DELAY_VFAT2_BIT2',
'TRIG_TAP_DELAY_VFAT2_BIT3',
'TRIG_TAP_DELAY_VFAT2_BIT4',
'TRIG_TAP_DELAY_VFAT2_BIT5',
'TRIG_TAP_DELAY_VFAT2_BIT6',
'TRIG_TAP_DELAY_VFAT2_BIT7',
'TRIG_TAP_DELAY_VFAT3_BIT0',
'TRIG_TAP_DELAY_VFAT3_BIT1',
'TRIG_TAP_DELAY_VFAT3_BIT2',
'TRIG_TAP_DELAY_VFAT3_BIT3',
'TRIG_TAP_DELAY_VFAT3_BIT4',
'TRIG_TAP_DELAY_VFAT3_BIT5',
'TRIG_TAP_DELAY_VFAT3_BIT6',
'TRIG_TAP_DELAY_VFAT3_BIT7',
'TRIG_TAP_DELAY_VFAT4_BIT0',
'TRIG_TAP_DELAY_VFAT4_BIT1',
'TRIG_TAP_DELAY_VFAT4_BIT2',
'TRIG_TAP_DELAY_VFAT4_BIT3',
'TRIG_TAP_DELAY_VFAT4_BIT4',
'TRIG_TAP_DELAY_VFAT4_BIT5',
'TRIG_TAP_DELAY_VFAT4_BIT6',
'TRIG_TAP_DELAY_VFAT4_BIT7',
'TRIG_TAP_DELAY_VFAT5_BIT0',
'TRIG_TAP_DELAY_VFAT5_BIT1',
'TRIG_TAP_DELAY_VFAT5_BIT2',
'TRIG_TAP_DELAY_VFAT5_BIT3',
'TRIG_TAP_DELAY_VFAT5_BIT4',
'TRIG_TAP_DELAY_VFAT5_BIT5',
'TRIG_TAP_DELAY_VFAT5_BIT6',
'TRIG_TAP_DELAY_VFAT5_BIT7',
'TRIG_TAP_DELAY_VFAT6_BIT0',
'TRIG_TAP_DELAY_VFAT6_BIT1',
'TRIG_TAP_DELAY_VFAT6_BIT2',
'TRIG_TAP_DELAY_VFAT6_BIT3',
'TRIG_TAP_DELAY_VFAT6_BIT4',
'TRIG_TAP_DELAY_VFAT6_BIT5',
'TRIG_TAP_DELAY_VFAT6_BIT6',
'TRIG_TAP_DELAY_VFAT6_BIT7',
'TRIG_TAP_DELAY_VFAT7_BIT0',
'TRIG_TAP_DELAY_VFAT7_BIT1',
'TRIG_TAP_DELAY_VFAT7_BIT2',
'TRIG_TAP_DELAY_VFAT7_BIT3',
'TRIG_TAP_DELAY_VFAT7_BIT4',
'TRIG_TAP_DELAY_VFAT7_BIT5',
'TRIG_TAP_DELAY_VFAT7_BIT6',
'TRIG_TAP_DELAY_VFAT7_BIT7',
'TRIG_TAP_DELAY_VFAT8_BIT0',
'TRIG_TAP_DELAY_VFAT8_BIT1',
'TRIG_TAP_DELAY_VFAT8_BIT2',
'TRIG_TAP_DELAY_VFAT8_BIT3',
'TRIG_TAP_DELAY_VFAT8_BIT4',
'TRIG_TAP_DELAY_VFAT8_BIT5',
'TRIG_TAP_DELAY_VFAT8_BIT6',
'TRIG_TAP_DELAY_VFAT8_BIT7',
'TRIG_TAP_DELAY_VFAT9_BIT0',
'TRIG_TAP_DELAY_VFAT9_BIT1',
'TRIG_TAP_DELAY_VFAT9_BIT2',
'TRIG_TAP_DELAY_VFAT9_BIT3',
'TRIG_TAP_DELAY_VFAT9_BIT4',
'TRIG_TAP_DELAY_VFAT9_BIT5',
'TRIG_TAP_DELAY_VFAT9_BIT6',
'TRIG_TAP_DELAY_VFAT9_BIT7',
'TRIG_TAP_DELAY_VFAT10_BIT0',
'TRIG_TAP_DELAY_VFAT10_BIT1',
'TRIG_TAP_DELAY_VFAT10_BIT2',
'TRIG_TAP_DELAY_VFAT10_BIT3',
'TRIG_TAP_DELAY_VFAT10_BIT4',
'TRIG_TAP_DELAY_VFAT10_BIT5',
'TRIG_TAP_DELAY_VFAT10_BIT6',
'TRIG_TAP_DELAY_VFAT10_BIT7',
'TRIG_TAP_DELAY_VFAT11_BIT0',
'TRIG_TAP_DELAY_VFAT11_BIT1',
'TRIG_TAP_DELAY_VFAT11_BIT2',
'TRIG_TAP_DELAY_VFAT11_BIT3',
'TRIG_TAP_DELAY_VFAT11_BIT4',
'TRIG_TAP_DELAY_VFAT11_BIT5',
'TRIG_TAP_DELAY_VFAT11_BIT6',
'TRIG_TAP_DELAY_VFAT11_BIT7',
'TRIG_TAP_DELAY_VFAT12_BIT0',
'TRIG_TAP_DELAY_VFAT12_BIT1',
'TRIG_TAP_DELAY_VFAT12_BIT2',
'TRIG_TAP_DELAY_VFAT12_BIT3',
'TRIG_TAP_DELAY_VFAT12_BIT4',
'TRIG_TAP_DELAY_VFAT12_BIT5',
'TRIG_TAP_DELAY_VFAT12_BIT6',
'TRIG_TAP_DELAY_VFAT12_BIT7',
'TRIG_TAP_DELAY_VFAT13_BIT0',
'TRIG_TAP_DELAY_VFAT13_BIT1',
'TRIG_TAP_DELAY_VFAT13_BIT2',
'TRIG_TAP_DELAY_VFAT13_BIT3',
'TRIG_TAP_DELAY_VFAT13_BIT4',
'TRIG_TAP_DELAY_VFAT13_BIT5',
'TRIG_TAP_DELAY_VFAT13_BIT6',
'TRIG_TAP_DELAY_VFAT13_BIT7',
'TRIG_TAP_DELAY_VFAT14_BIT0',
'TRIG_TAP_DELAY_VFAT14_BIT1',
'TRIG_TAP_DELAY_VFAT14_BIT2',
'TRIG_TAP_DELAY_VFAT14_BIT3',
'TRIG_TAP_DELAY_VFAT14_BIT4',
'TRIG_TAP_DELAY_VFAT14_BIT5',
'TRIG_TAP_DELAY_VFAT14_BIT6',
'TRIG_TAP_DELAY_VFAT14_BIT7',
'TRIG_TAP_DELAY_VFAT15_BIT0',
'TRIG_TAP_DELAY_VFAT15_BIT1',
'TRIG_TAP_DELAY_VFAT15_BIT2',
'TRIG_TAP_DELAY_VFAT15_BIT3',
'TRIG_TAP_DELAY_VFAT15_BIT4',
'TRIG_TAP_DELAY_VFAT15_BIT5',
'TRIG_TAP_DELAY_VFAT15_BIT6',
'TRIG_TAP_DELAY_VFAT15_BIT7',
'TRIG_TAP_DELAY_VFAT16_BIT0',
'TRIG_TAP_DELAY_VFAT16_BIT1',
'TRIG_TAP_DELAY_VFAT16_BIT2',
'TRIG_TAP_DELAY_VFAT16_BIT3',
'TRIG_TAP_DELAY_VFAT16_BIT4',
'TRIG_TAP_DELAY_VFAT16_BIT5',
'TRIG_TAP_DELAY_VFAT16_BIT6',
'TRIG_TAP_DELAY_VFAT16_BIT7',
'TRIG_TAP_DELAY_VFAT17_BIT0',
'TRIG_TAP_DELAY_VFAT17_BIT1',
'TRIG_TAP_DELAY_VFAT17_BIT2',
'TRIG_TAP_DELAY_VFAT17_BIT3',
'TRIG_TAP_DELAY_VFAT17_BIT4',
'TRIG_TAP_DELAY_VFAT17_BIT5',
'TRIG_TAP_DELAY_VFAT17_BIT6',
'TRIG_TAP_DELAY_VFAT17_BIT7',
'TRIG_TAP_DELAY_VFAT18_BIT0',
'TRIG_TAP_DELAY_VFAT18_BIT1',
'TRIG_TAP_DELAY_VFAT18_BIT2',
'TRIG_TAP_DELAY_VFAT18_BIT3',
'TRIG_TAP_DELAY_VFAT18_BIT4',
'TRIG_TAP_DELAY_VFAT18_BIT5',
'TRIG_TAP_DELAY_VFAT18_BIT6',
'TRIG_TAP_DELAY_VFAT18_BIT7',
'TRIG_TAP_DELAY_VFAT19_BIT0',
'TRIG_TAP_DELAY_VFAT19_BIT1',
'TRIG_TAP_DELAY_VFAT19_BIT2',
'TRIG_TAP_DELAY_VFAT19_BIT3',
'TRIG_TAP_DELAY_VFAT19_BIT4',
'TRIG_TAP_DELAY_VFAT19_BIT5',
'TRIG_TAP_DELAY_VFAT19_BIT6',
'TRIG_TAP_DELAY_VFAT19_BIT7',
'TRIG_TAP_DELAY_VFAT20_BIT0',
'TRIG_TAP_DELAY_VFAT20_BIT1',
'TRIG_TAP_DELAY_VFAT20_BIT2',
'TRIG_TAP_DELAY_VFAT20_BIT3',
'TRIG_TAP_DELAY_VFAT20_BIT4',
'TRIG_TAP_DELAY_VFAT20_BIT5',
'TRIG_TAP_DELAY_VFAT20_BIT6',
'TRIG_TAP_DELAY_VFAT20_BIT7',
'TRIG_TAP_DELAY_VFAT21_BIT0',
'TRIG_TAP_DELAY_VFAT21_BIT1',
'TRIG_TAP_DELAY_VFAT21_BIT2',
'TRIG_TAP_DELAY_VFAT21_BIT3',
'TRIG_TAP_DELAY_VFAT21_BIT4',
'TRIG_TAP_DELAY_VFAT21_BIT5',
'TRIG_TAP_DELAY_VFAT21_BIT6',
'TRIG_TAP_DELAY_VFAT21_BIT7',
'TRIG_TAP_DELAY_VFAT22_BIT0',
'TRIG_TAP_DELAY_VFAT22_BIT1',
'TRIG_TAP_DELAY_VFAT22_BIT2',
'TRIG_TAP_DELAY_VFAT22_BIT3',
'TRIG_TAP_DELAY_VFAT22_BIT4',
'TRIG_TAP_DELAY_VFAT22_BIT5',
'TRIG_TAP_DELAY_VFAT22_BIT6',
'TRIG_TAP_DELAY_VFAT22_BIT7',
'TRIG_TAP_DELAY_VFAT23_BIT0',
'TRIG_TAP_DELAY_VFAT23_BIT1',
'TRIG_TAP_DELAY_VFAT23_BIT2',
'TRIG_TAP_DELAY_VFAT23_BIT3',
'TRIG_TAP_DELAY_VFAT23_BIT4',
'TRIG_TAP_DELAY_VFAT23_BIT5',
'TRIG_TAP_DELAY_VFAT23_BIT6',
'TRIG_TAP_DELAY_VFAT23_BIT7']
```


#### `VFAT2_CHANNEL_CONF`
```
Columns for GEM_VFAT2_CHANNEL_CONF
Executing query: select * from CMS_GEM_MUON_COND.GEM_VFAT2_CHANNEL_CONF WHERE 1 = 0
['RECORD_ID',
'CONDITION_DATA_SET_ID',
'CHANNEL_REG_0',
'CHANNEL_REG_1',
'CHANNEL_REG_2',
'CHANNEL_REG_3',
'CHANNEL_REG_4',
'CHANNEL_REG_5',
'CHANNEL_REG_6',
'CHANNEL_REG_7',
'CHANNEL_REG_8',
'CHANNEL_REG_9',
'CHANNEL_REG_10',
'CHANNEL_REG_11',
'CHANNEL_REG_12',
'CHANNEL_REG_13',
'CHANNEL_REG_14',
'CHANNEL_REG_15',
'CHANNEL_REG_16',
'CHANNEL_REG_17',
'CHANNEL_REG_18',
'CHANNEL_REG_19',
'CHANNEL_REG_20',
'CHANNEL_REG_21',
'CHANNEL_REG_22',
'CHANNEL_REG_23',
'CHANNEL_REG_24',
'CHANNEL_REG_25',
'CHANNEL_REG_26',
'CHANNEL_REG_27',
'CHANNEL_REG_28',
'CHANNEL_REG_29',
'CHANNEL_REG_30',
'CHANNEL_REG_31',
'CHANNEL_REG_32',
'CHANNEL_REG_33',
'CHANNEL_REG_34',
'CHANNEL_REG_35',
'CHANNEL_REG_36',
'CHANNEL_REG_37',
'CHANNEL_REG_38',
'CHANNEL_REG_39',
'CHANNEL_REG_40',
'CHANNEL_REG_41',
'CHANNEL_REG_42',
'CHANNEL_REG_43',
'CHANNEL_REG_44',
'CHANNEL_REG_45',
'CHANNEL_REG_46',
'CHANNEL_REG_47',
'CHANNEL_REG_48',
'CHANNEL_REG_49',
'CHANNEL_REG_50',
'CHANNEL_REG_51',
'CHANNEL_REG_52',
'CHANNEL_REG_53',
'CHANNEL_REG_54',
'CHANNEL_REG_55',
'CHANNEL_REG_56',
'CHANNEL_REG_57',
'CHANNEL_REG_58',
'CHANNEL_REG_59',
'CHANNEL_REG_60',
'CHANNEL_REG_61',
'CHANNEL_REG_62',
'CHANNEL_REG_63',
'CHANNEL_REG_64',
'CHANNEL_REG_65',
'CHANNEL_REG_66',
'CHANNEL_REG_67',
'CHANNEL_REG_68',
'CHANNEL_REG_69',
'CHANNEL_REG_70',
'CHANNEL_REG_71',
'CHANNEL_REG_72',
'CHANNEL_REG_73',
'CHANNEL_REG_74',
'CHANNEL_REG_75',
'CHANNEL_REG_76',
'CHANNEL_REG_77',
'CHANNEL_REG_78',
'CHANNEL_REG_79',
'CHANNEL_REG_80',
'CHANNEL_REG_81',
'CHANNEL_REG_82',
'CHANNEL_REG_83',
'CHANNEL_REG_84',
'CHANNEL_REG_85',
'CHANNEL_REG_86',
'CHANNEL_REG_87',
'CHANNEL_REG_88',
'CHANNEL_REG_89',
'CHANNEL_REG_90',
'CHANNEL_REG_91',
'CHANNEL_REG_92',
'CHANNEL_REG_93',
'CHANNEL_REG_94',
'CHANNEL_REG_95',
'CHANNEL_REG_96',
'CHANNEL_REG_97',
'CHANNEL_REG_98',
'CHANNEL_REG_99',
'CHANNEL_REG_100',
'CHANNEL_REG_101',
'CHANNEL_REG_102',
'CHANNEL_REG_103',
'CHANNEL_REG_104',
'CHANNEL_REG_105',
'CHANNEL_REG_106',
'CHANNEL_REG_107',
'CHANNEL_REG_108',
'CHANNEL_REG_109',
'CHANNEL_REG_110',
'CHANNEL_REG_111',
'CHANNEL_REG_112',
'CHANNEL_REG_113',
'CHANNEL_REG_114',
'CHANNEL_REG_115',
'CHANNEL_REG_116',
'CHANNEL_REG_117',
'CHANNEL_REG_118',
'CHANNEL_REG_119',
'CHANNEL_REG_120',
'CHANNEL_REG_121',
'CHANNEL_REG_122',
'CHANNEL_REG_123',
'CHANNEL_REG_124',
'CHANNEL_REG_125',
'CHANNEL_REG_126',
'CHANNEL_REG_127']
```

#### `VFAT2_CHIP_CONF`
```
Columns for GEM_VFAT2_CHIP_CONF
Executing query: select * from CMS_GEM_MUON_COND.GEM_VFAT2_CHIP_CONF WHERE 1 = 0
['RECORD_ID',
'CONDITION_DATA_SET_ID',
'CTRL_REG0',
'CTRL_REG1',
'CTRL_REG2',
'CTRL_REG3',
'PREAMP_IN_BIAS_CTRL',
'PREAMP_OUT_BIAS_CTRL',
'PREAMP_FEED_BIAS_CTRL',
'SHAPER_BIAS_CTRL',
'SHAPER_FEED_BIAS_CTRL',
'COMP_BIAS_CTRL',
'LATENCY_CTRL',
'VTHRESHOLD_CTRL1',
'VTHRESHOLD_CTRL2',
'VCAL_CTRL',
'CALPHASE_CTRL',
'HW_RW_REG']
```

#### `VFAT3_CHANNEL_CONF`
```
Columns for GEM_VFAT3_CHANNEL_CONF
Executing query: select * from CMS_GEM_MUON_COND.GEM_VFAT3_CHANNEL_CONF WHERE 1 = 0
['RECORD_ID',
'CONDITION_DATA_SET_ID',
'CFG_CHANNEL_0',
'CFG_CHANNEL_1',
'CFG_CHANNEL_2',
'CFG_CHANNEL_3',
'CFG_CHANNEL_4',
'CFG_CHANNEL_5',
'CFG_CHANNEL_6',
'CFG_CHANNEL_7',
'CFG_CHANNEL_8',
'CFG_CHANNEL_9',
'CFG_CHANNEL_10',
'CFG_CHANNEL_11',
'CFG_CHANNEL_12',
'CFG_CHANNEL_13',
'CFG_CHANNEL_14',
'CFG_CHANNEL_15',
'CFG_CHANNEL_16',
'CFG_CHANNEL_17',
'CFG_CHANNEL_18',
'CFG_CHANNEL_19',
'CFG_CHANNEL_20',
'CFG_CHANNEL_21',
'CFG_CHANNEL_22',
'CFG_CHANNEL_23',
'CFG_CHANNEL_24',
'CFG_CHANNEL_25',
'CFG_CHANNEL_26',
'CFG_CHANNEL_27',
'CFG_CHANNEL_28',
'CFG_CHANNEL_29',
'CFG_CHANNEL_30',
'CFG_CHANNEL_31',
'CFG_CHANNEL_32',
'CFG_CHANNEL_33',
'CFG_CHANNEL_34',
'CFG_CHANNEL_35',
'CFG_CHANNEL_36',
'CFG_CHANNEL_37',
'CFG_CHANNEL_38',
'CFG_CHANNEL_39',
'CFG_CHANNEL_40',
'CFG_CHANNEL_41',
'CFG_CHANNEL_42',
'CFG_CHANNEL_43',
'CFG_CHANNEL_44',
'CFG_CHANNEL_45',
'CFG_CHANNEL_46',
'CFG_CHANNEL_47',
'CFG_CHANNEL_48',
'CFG_CHANNEL_49',
'CFG_CHANNEL_50',
'CFG_CHANNEL_51',
'CFG_CHANNEL_52',
'CFG_CHANNEL_53',
'CFG_CHANNEL_54',
'CFG_CHANNEL_55',
'CFG_CHANNEL_56',
'CFG_CHANNEL_57',
'CFG_CHANNEL_58',
'CFG_CHANNEL_59',
'CFG_CHANNEL_60',
'CFG_CHANNEL_61',
'CFG_CHANNEL_62',
'CFG_CHANNEL_63',
'CFG_CHANNEL_64',
'CFG_CHANNEL_65',
'CFG_CHANNEL_66',
'CFG_CHANNEL_67',
'CFG_CHANNEL_68',
'CFG_CHANNEL_69',
'CFG_CHANNEL_70',
'CFG_CHANNEL_71',
'CFG_CHANNEL_72',
'CFG_CHANNEL_73',
'CFG_CHANNEL_74',
'CFG_CHANNEL_75',
'CFG_CHANNEL_76',
'CFG_CHANNEL_77',
'CFG_CHANNEL_78',
'CFG_CHANNEL_79',
'CFG_CHANNEL_80',
'CFG_CHANNEL_81',
'CFG_CHANNEL_82',
'CFG_CHANNEL_83',
'CFG_CHANNEL_84',
'CFG_CHANNEL_85',
'CFG_CHANNEL_86',
'CFG_CHANNEL_87',
'CFG_CHANNEL_88',
'CFG_CHANNEL_89',
'CFG_CHANNEL_90',
'CFG_CHANNEL_91',
'CFG_CHANNEL_92',
'CFG_CHANNEL_93',
'CFG_CHANNEL_94',
'CFG_CHANNEL_95',
'CFG_CHANNEL_96',
'CFG_CHANNEL_97',
'CFG_CHANNEL_98',
'CFG_CHANNEL_99',
'CFG_CHANNEL_100',
'CFG_CHANNEL_101',
'CFG_CHANNEL_102',
'CFG_CHANNEL_103',
'CFG_CHANNEL_104',
'CFG_CHANNEL_105',
'CFG_CHANNEL_106',
'CFG_CHANNEL_107',
'CFG_CHANNEL_108',
'CFG_CHANNEL_109',
'CFG_CHANNEL_110',
'CFG_CHANNEL_111',
'CFG_CHANNEL_112',
'CFG_CHANNEL_113',
'CFG_CHANNEL_114',
'CFG_CHANNEL_115',
'CFG_CHANNEL_116',
'CFG_CHANNEL_117',
'CFG_CHANNEL_118',
'CFG_CHANNEL_119',
'CFG_CHANNEL_120',
'CFG_CHANNEL_121',
'CFG_CHANNEL_122',
'CFG_CHANNEL_123',
'CFG_CHANNEL_124',
'CFG_CHANNEL_125',
'CFG_CHANNEL_126',
'CFG_CHANNEL_127']
```

#### `VFAT3_CHIP_CONF`
```
Columns for GEM_VFAT3_CHIP_CONF
Executing query: select * from CMS_GEM_MUON_COND.GEM_VFAT3_CHIP_CONF WHERE 1 = 0
['RECORD_ID',
'CONDITION_DATA_SET_ID',
'GLB_LOGIC_CTRL',
'GLB_DATA_CTRL',
'GLB_FE_CTRL',
'GLB_CFD_CTRL',
'GLB_MON_CTRL',
'GLB_IREF_CTRL',
'GLB_THRESHOLD_CTRL',
'GLB_HYSTERISIS_CTRL',
'GLB_LATENCY_CTRL',
'GLB_CAL0_CFG_CTRL',
'GLB_CAL1_CFG_CTRL',
'GLB_CFD_BIAS_CTRL',
'GLB_PREAMP_BIAS_CTRL0',
'GLB_PREAMP_BIAS_CTRL1',
'GLB_SHAPER_BIAS_CTRL0',
'GLB_SHAPER_BIAS_CTRL1',
'GLB_SD_BIAS_CTRL',
'HW_RW_REG']
```

lmoureaux commented:
> Just trying to get to what is implemented here:
> Is this a good executive summary?

It is. Let me expand it a bit **in bold face**:

1. **Objects representing VFAT3 chip and channel configuration records (both DB and XML)**
1. Object representing XML specification of **any** configuration record **(templated to the class for item 1)**
1. methods to access the individual configuration parameters/properties
1. xerces parsing **and synthesizing** back end to make aforementioned functionality work
1. **Tests to make sure that everything works as expected**
